akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphPartitionSpec.scala

```diff
@@ -0,0 +1,173 @@
+          case _                   ⇒ 1
+        }))
+        Source(List("this", "is", "just", "another", "test")) ~> partition.in
+        partition.out(0) ~> Sink.fromSubscriber(c1)
+        partition.out(1) ~> Sink.fromSubscriber(c2)
+        ClosedShape
+      }).run()
+
+      c1.request(1)
+      c2.request(4)
+      c1.expectNext("another")
+      c2.expectNext("this")
+      c2.expectNext("is")
+      c2.expectNext("just")
+      c2.expectNext("test")
+
+    }
+
+    "remember first pull even though first element targeted another out" in assertAllStagesStopped {
+      val c1 = TestSubscriber.probe[Int]()
+      val c2 = TestSubscriber.probe[Int]()
+
+      RunnableGraph.fromGraph(GraphDSL.create() { implicit b ⇒
+        val partition = b.add(Partition[Int](2, { case l if l < 6 ⇒ 0; case _ ⇒ 1 }))
+        Source(List(6, 3)) ~> partition.in
+        partition.out(0) ~> Sink.fromSubscriber(c1)
+        partition.out(1) ~> Sink.fromSubscriber(c2)
+        ClosedShape
+      }).run()
+
+      c1.request(1)
+      c1.expectNoMsg(1.seconds)
+      c2.request(1)
+      c2.expectNext(6)
+      c1.expectNext(3)
+    }
+
+    "cancel upstream when downstreams cancel" in assertAllStagesStopped {
+      val p1 = TestPublisher.probe[Int]()
+      val c1 = TestSubscriber.probe[Int]()
+      val c2 = TestSubscriber.probe[Int]()
+
+      RunnableGraph.fromGraph(GraphDSL.create() { implicit b ⇒
+        val partition = b.add(Partition[Int](2, { case l if l < 6 ⇒ 0; case _ ⇒ 1 }))
+        Source.fromPublisher(p1.getPublisher) ~> partition.in
+        partition.out(0) ~> Flow[Int].buffer(16, OverflowStrategy.backpressure) ~> Sink.fromSubscriber(c1)
+        partition.out(1) ~> Flow[Int].buffer(16, OverflowStrategy.backpressure) ~> Sink.fromSubscriber(c2)
+        ClosedShape
+      }).run()
+
+      val p1Sub = p1.expectSubscription()
+      val sub1 = c1.expectSubscription()
+      val sub2 = c2.expectSubscription()
+      sub1.request(3)
+      sub2.request(3)
+      p1Sub.sendNext(1)
+      p1Sub.sendNext(8)
+      c1.expectNext(1)
+      c2.expectNext(8)
+      p1Sub.sendNext(2)
+      c1.expectNext(2)
+      sub1.cancel()
+      sub2.cancel()
+      p1Sub.expectCancellation()
+    }
+
+    "work with merge" in assertAllStagesStopped {
+      val c1 = TestSubscriber.probe[Int]()
+      val c2 = TestSubscriber.probe[Int]()
+      val probe = TestSubscriber.probe[Int]()
+      val t = TestSink.probe[Int]
+
+      RunnableGraph.fromGraph(GraphDSL.create() { implicit b ⇒
+        val partition = b.add(Partition[Int](2, { case l if l < 4 ⇒ 0; case _ ⇒ 1 }))
+        val merge = b.add(Merge[Int](2))
+        Source(List(5, 2, 9, 1, 1, 1, 10)) ~> partition.in
+        partition.out(0) ~> merge.in(0)
+        partition.out(1) ~> merge.in(1)
+        merge.out ~> Sink.fromSubscriber(probe)
+
+        ClosedShape
+      }).run()
+
+      val subscription = probe.expectSubscription()
+
+      var collected = Set.empty[Int]
+      for (_ ← 1 to 7) {
```

Comment:
No need for this, there is also a version of expect that takes a collection, or even better, drain it with a Sink.seq